### ISSUE TITLE
NLL_gen loss with initialization

### ISSUE THREAD

**Issue by lethaiq** [weilinie/RelGAN#10]
@weilinie ,

Thank you for the implementation. I wonder if you always observe the same behavior of NLL_gen loss curve with different initialization? I am not observing the same behavior on another code base so I am wondering what I am doing wrong.

<img width="566" alt="image" src="https://user-images.githubusercontent.com/13818722/70863010-68b7c300-1f11-11ea-9383-80634a604fab.png">

Thanks!

**Comment by weilinie** [comment]
Thanks for your interests in this work. In terms of NLL_gen, there always exists a tendency: it first increases and then decreases from my observations though variance also exists due to the randomness. I’m not quite sure what other different behavior you have observed, but one thing you may check is if the pre-training stage has reached a similar NLL_gen score as shown in the above figure to make sure the pre-training does not end up with a bad initialization.

**Comment by lethaiq** [comment]
> Thanks for your interests in this work. In terms of NLL_gen, there always exists a tendency: it first increases and then decreases from my observations though variance also exists due to the randomness. I’m not quite sure what other different behavior you have observed, but one thing you may check is if the pre-training stage has reached a similar NLL_gen score as shown in the above figure to make sure the pre-training does not end up with a bad initialization.

Thanks @weilinie. From your experience on other datasets rather than COO image, what would be a good target for NLL_gen score during pre-training? Will there be a big difference during adversarial training with different final NLL_gen score during pre-training?

**Comment by weilinie** [comment]
Usually a lower NLL_gen in the end of pre-training is better for adversarial training. From my observations, different final NLL_gen scores during pre-training may result in different final NLL_gen scores during adversarial training. However, the quality-diversity tradeoff during adversarial training seems to always exist.

**Comment by lethaiq** [comment]
> Usually a lower NLL_gen in the end of pre-training is better for adversarial training. From my observations, different final NLL_gen scores during pre-training may result in different final NLL_gen scores during adversarial training. However, the quality-diversity tradeoff during adversarial training seems to always exist.

Thanks @weilinie for very helpful observations.

**Comment by lethaiq** [comment]
@weilinie I have one more question. Why didn't you use RMC_att model but use RMC_vanilla instead? I wonder if there is any specific reasons why you decide to do so, or RMC_att does not work well under the current GAN settings?

**Comment by weilinie** [comment]
I indeed tried different classifier networks as the discriminator but RMC_vanilla works the best. Yes, RMC_att didn’t work well in current GAN settings.

**Comment by lethaiq** [comment]
> I indeed tried different classifier networks as the discriminator but RMC_vanilla works the best. Yes, RMC_att didn’t work well in current GAN settings.

Thank you so much for your response.

**Comment by lethaiq** [comment]
@weilinie I have another question, I am testing with emnlp dataset, which has a much longer sequence length than COO image dataset, and seems like the NLL_gen will not reach the same level as of COO image dataset after MLE pre-training. Since in the paper we do not have the graph for emnlp dataset, should I expect a very similar behavior with COO image dataset? My code might be wrong somewhere...

**Comment by weilinie** [comment]
In my experiments on EMNLP, there may be some fluctuations right in the beginning stage of adv-training, but we can still see the diversity-quality tradeoff afterwards.

**Comment by lethaiq** [comment]
> In my experiments on EMNLP, there may be some fluctuations right in the beginning stage of adv-training, but we can still see the diversity-quality tradeoff afterwards.

Thanks @weilinie. I cannot reach NLL_gen for EMNLP of around 0.55 like COO Image during pre-training, is it expected behavior?

**Comment by weilinie** [comment]
Yes. If I remember correctly, the NLL_gen for EMNLP after MLE pre-training is around 2.

**Comment by lethaiq** [comment]
> Yes. If I remember correctly, the NLL_gen for EMNLP after MLE pre-training is around 2.

@weilinie Thank you very much. So from our discussion, the general image for a "good" training behaviors are:
(i) MLE pre-training of generator improves (NLL_gen should decrease to a certain point)
(ii) Then Adversarial training will continue and show the quality-diversity trade-off.

In the paper, I am pretty impressed when you show results on COO image that the NLL_gen of RelGAN is even better than MLE (Page 7). Is this NLL_gen of MLE the one recorded the end of MLE pre-training, or is this MLE recorded for a separate MLE training with longer iterations?

**Comment by weilinie** [comment]
The MLE model reported in the tables is LSTM-based generator. This is why the difference happens as you observed.

**Comment by lethaiq** [comment]
> The MLE model reported in the tables is LSTM-based generator. This is why the difference happens as you observed.

Thanks @weilinie this makes sense now. Thank you for spending your time to answer my questions, very helpful and informative!
